tests/server/master_server/chunk_server_manager_test.cc

```diff
@@ -17,48 +20,637 @@ class ChunkServerManagerTest : public ::testing::Test {
+        std::async(std::launch::async, [&, i]() {
+          return ChunkServerManager::GetInstance().AllocateChunkServer(
+              std::to_string(i),
+              /*server_count=*/1);
+        }));
+
+    // we only want to create 5 more chunkservers
+    if (i < concurrently_registered_servers_count) {
+      const ushort server_id = preregistered_servers_count + i;
+
+      server_registration_thread_futures.push_back(
+          std::async(std::launch::async, [&, server_id]() {
+            protos::ChunkServerLocation* server_location =
+                new protos::ChunkServerLocation(
+                    std::move(CreateChunkServerLocation(
+                        std::string("192.168.1.") + std::to_string(server_id),
+                        server_id)));
+
+            std::shared_ptr<protos::ChunkServer> chunk_server(
+                new protos::ChunkServer());
+            chunk_server->set_allocated_location(server_location);
+            // Giving them 100 to 109mb, can only store 1 chunk
+            chunk_server->set_available_disk_mb(100 + server_id);
+
+            ChunkServerManager::GetInstance().RegisterChunkServer(chunk_server);
+
+            // store all the created locations.
+            // Doing this here, since the set synchronizes threads and don't
+            // want that before we register.
+            created_locations.insert(*server_location);
+          }));
+    }
+  }
+
+  // Wait for each server allocation thread future and get result.
+  // No need to wait for the chunkserver registration threads.
+  for (ushort i = 0; i < server_allocation_thread_count; i++) {
+    auto allocated_locations = server_allocation_thread_futures[i].get();
+
+    // We requested for 1 server and every chunk should get 1 since we added 5
+    // more to the preregistered 5.
+    EXPECT_EQ(1, allocated_locations.size());
+
+    // Get the location
+    auto location = *(allocated_locations.begin());
+
+    EXPECT_TRUE(created_locations.contains(location));
+
+    // Remove to check that this location was only assigned to this thread's
+    // chunk, since we configured each chunkserver with disk for just 1 chunk.
+    // If we see this location then the check above will fail.
+    created_locations.erase(location);
+
+    // Ensure that GetChunkLocation returns this location
+    auto chunk_locations =
+        ChunkServerManager::GetInstance().GetChunkLocations(std::to_string(i));
+
+    EXPECT_EQ(1, chunk_locations.size());
+
+    EXPECT_EQ(location, *(chunk_locations.begin()));
+  }
+}
+
+// Test concurrent allocation of chunk servers while also concurrently updating
+// chunkservers.
+TEST_F(ChunkServerManagerTest, ConcurrentAllocateAndUpdateChunkServers) {
+  const ushort chunk_servers_count = 10;
+  ChunkServerLocationFlatSet created_locations;
+
+  // Register 10 chunkservers with chunks 101 to 110
+  for (ushort i = 1; i <= chunk_servers_count; i++) {
+    protos::ChunkServerLocation* server_location =
```

Comment:
I guess there is a minor resource leak here? probably ok for testing.

Reply:
nope. Doing this below      chunk_server->set_allocated_location(server_location);  the chunk_server takes control of the allocated memory and will free on destruction.